src/arch/x86/64/traps.S

```diff
@@ -679,6 +688,47 @@ BEGIN_FUNC(handle_vmexit)
     push    %rbx
     push    %rax
 
+    # Save the guest's GS, Shadow GS, and FS
+    mov $0xC0000100, %ecx
+    rdmsr
+    shl $0x20,%rdx
+    or %rdx, %rax
+    pushq %rax # Save FS
+    swapgs
+    mov $0xC0000101, %ecx
+    rdmsr
+    shl $0x20,%rdx
+    or %rdx, %rax
+    pushq %rax # Save Shadow GS
+    swapgs
+    rdmsr
+    shl $0x20,%rdx
+    or %rdx, %rax
+    pushq %rax # Save GS
+
+#ifdef ENABLE_SMP_SUPPORT
+
+    # RSP now at host_gs registers
+    sub $0x18, %rsp
+
+    # Restore the host's GS and Shadow GS
```

Comment:
Why is the restoring host's GS and shadow GS guarded by ENABLE_SMP_SUPPORT? What does the shadow GS mean? Kernel GS or User GS?

Reply:
So I wrote this code ~2.5 years ago, so I'll do my best to explain some decisions which may or may not make any sense.

I believe the existing `handle_vmexit` had `swapgs` wrapped around SMP support. So I put all the GS restoration around the same wrapper. I believe i've tested this without that wrapper and everything functions properly. I even made commit that Damon may have merged in that always performs the GS save/restore

Reply:
As for Shadow GS, its the MSR value at `0xc0000102`, and I believe that is what Linux refers to it as. seL4 uses it for the `node_info` struct. Since the guest and seL4 use this value, we have to make sure to save/restore both versions, depending on which piece of software is currently executing. (That might explain why the host GS and FS are wrapped around SMP support, and hints that there is an optimal way of handling this)

```
x86_wrmsr(IA32_KERNEL_GS_BASE_MSR, (word_t)&node_info[cpu_index]);
```

Reply:
If I understand correctly, the code above is trying to save the IA32_GS_BASE, IA32_KERNEL_GS_BASE and IA32_FS_BASE onto the stack, right? So, is it possible to just use the rdmsr with MSR numbers directly without  using "SWAPGS"? Therefore, the following restore code could also be avoided?